src/rewriter/README.md

```diff
@@ -0,0 +1,231 @@
+# The Grammar of RARE
+
+This document describe the syntax of the RARE language, which stands for
+**r**ewrites, **a**utomatically **re**constructed.
+
+``` dsl
+   <rule> ::= (define-rule       <symbol> (<par>*) [<defs>]        <expr> <expr>)
+            | (define-rule*      <symbol> (<par>*) [<defs>]        <expr> <expr> [<expr>])
+            | (define-cond-rule  <symbol> (<par>*) [<defs>] <expr> <expr> <expr>)
+    <par> ::= (<symbol> <sort> [<attr>])
+   <sort> ::= ? | <symbol> | ?<symbol> | (<symbol> <sort>+) | (<symbol> <idx>+)
+    <idx> ::= <numeral>
+   <attr> ::= :list
+   <expr> ::= <const> | <id> | (<id> <expr>+)
+     <id> ::= <symbol> | (_ <id> <expr>+)
+<binding> ::= (<symbol> <expr>)
+   <defs> ::= (def <binding>+)
+```
+
+## Rules
+
+### Matching
+
+The basic form of a RARE rewrite rule uses `define-rule`. Each rare rule has a
+parameter list. Variables appearing in the parameter list can match any
+expression of the same type.
+
+The two expressions in the body of `define-rule` indicate the *match* and
+*target* of the rewrites. The match and target use SMT-LIB3 syntax.
+
+``` lisp
+(define-rule substr-empty
+	((m Int) (n Int))
+	(str.substr "" m n)
+	"")
+```
+
+In this case, the match is `(str.substr "" m n)` and the target is `""`. For
+example, when the rule is applied to `(str.substr "" 3 5)`, the parameter list
+variables `m,n` are instantiated to `3,5`, respectively, and the expression is
+rewritten as `""`.
+
+The match expression is purely syntactic and only syntactically identical terms
+can be matched.  For example `(= (str.++ x1 x2) x2)`matches `a ++ b = b` but not
+`a ++ b = c`.
+
+### Definition List
+
+An optional *definition list* may appear immediately after the parameter list to
+include variables representing expressions.
+
+The definition list, indicated by `def`, allow the rewrite rule to be more
+succinctly expressed and readable. Each variable in the definition list is
+replaced by its expression. `def` can be viewed as a special case of `let` which
+only exists in the outermost layer.
```

Comment:
Since `let` was removed, maybe don' t mention it?

Reply:
This is only here to give people an intuition of what `def` does since let-in constructs exist in many languages.